src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ServiceChannelFactory.cs

```diff
@@ -390,6 +395,33 @@ internal static ServiceChannel GetServiceChannel(object transparentProxy)
                 return null;
         }
 
+        private async Task OnCloseAsyncInternal(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            while (true)
+            {
+                int count;
+                IChannel channel;
+                lock (ThisLock)
+                {
+                    count = _channelsList.Count;
+                    if (count == 0)
+                        return;
+                    channel = _channelsList[0];
+                }
+
+                IAsyncCommunicationObject asyncChannel = channel as IAsyncCommunicationObject;
+                if (asyncChannel != null)
+                {
+                    await asyncChannel.CloseAsync(timeoutHelper.RemainingTime());
+                }
+                else
+                {
+                    await Task.Factory.FromAsync(channel.BeginClose, channel.EndClose, TaskCreationOptions.None);
```

Comment:
Reviewers: currently our channels are always IAsyncCommunicationObject and we take the nice call path above.  This else clause is to deal with the possibility we encounter a channel that is not (I don't know if that is possible). Thoughts on whether we should assert here?  Notice we don't pass the timeout through in this else case -- I couldn't find a way to do that with existing logic, but did not pursue deeply because I think this code path may not be needed.